### ISSUE TITLE
Support enabling policy when a NetworkPolicy selects a pod

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#788]
- Extend policy flag to daemon to have three modes
  - default (non k8s: enabled if at least one policy exists, k8s: enabled on pod which are selected)
  - enabled (always on)
  - disabled (always off)
- Add field to `endpoint list` to indicate whether policy is enabled

**Comment by ianvernon** [comment]
Desired output to `endpoint list`
```
cilium endpoint list
ENDPOINT  POLICY   IDENTITY   LABELS (source:key[=value])                   IPv6                     IPv4          STATUS   
33115        YES     2323     k8s:io.kubernetes.pod.namespace=kube-system   fd02::c0a8:210b:0:815b   10.11.220.6   ready    
                              k8s:k8s-app=kube-dns
33115         NO     2323     k8s:io.kubernetes.pod.namespace=kube-system   fd02::c0a8:210b:0:815b   10.11.220.6   ready    
                              k8s:k8s-app=kube-dns
```

**Comment by ianvernon** [comment]
I did the following to test whether the enabling / disabling code works within the build VM (not K8's enabled, but to test the behavior for policies that are added by K8's):
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ docker run -dt --net=$TEST_NET --name foo -l id.foo -l id.teamA tgraf/netperf && docker run -dt --net=$TEST_NET --name bar -l id.bar -l id.teamA tgraf/netperf && docker run -dt --net=$TEST_NET --name baz -l id.baz tgra
f/netperf
8c5954086a999b5e0185b91f44bffd4f5418a33b97a463076a028d92741b2365
ec35211bde369550d464d56425fd4416cf051c8ccdb8ad0ff16f0f1bcd121d14
05c0fbe34e081d34b0f93ff22513cc69b16b974aa34c694de4c08b2bedbee328
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY   IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
56326      false    275        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      false    274        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                               cilium:id.teamA                                                                 
63464      false    276        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                               cilium:id.teamA                                                                 
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cat <<EOF | cilium -D policy import -
[{
    "endpointSelector": {"matchLabels":{"id.bar":""}},
    "ingress": [{
        "fromEndpoints": [
            {"matchLabels":{"reserved:host":""}},
            {"matchLabels":{"id.foo":""}}
        ]
    }]
}]
EOF
DEBU[0000] Entering directory -...                      
DEBU[0000] Loading file -                               
DEBU[0000] Constructed policy object for import [0xc420288600] 
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY   IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
56326      false    275        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      true     274        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                               cilium:id.teamA                                                                 
63464      false    276        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                               cilium:id.teamA                                                                 
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ vi /var/log/syslog
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium policy delete --all
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium endpoint list
ENDPOINT   POLICY   IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
56326      false    275        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      false    274        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                               cilium:id.teamA                                                                 
63464      false    276        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                               cilium:id.teamA                                                  
```

@tgraf @aanm can you please confirm that this is the correct behavior we should see in the CLI?

**Comment by aanm** [comment]
that looks great.

I'm not sure about the `false`/`true` in `POLICY`. Maybe change it to something else like `isolated`/`open`?

**Comment by tgraf** [comment]
> I'm not sure about the false/true in POLICY. Maybe change it to something else like isolated/open?

My vote is on `enabled` / `disabled`

**Comment by aanm** [comment]
> My vote is on enabled / disabled

Couldn't this be confuse if we actually have the endpoint's `Policy` configuration showing `enabled` and then with the `cilium endpoint list` we would show disabled? It's not the policy per se that is enabled or disabled, it's the policy enforcement. Perhaps we could change the column name to `ENFORCED` or something like it?

**Comment by tgraf** [comment]
Or `POLICY` with `Yes/No`? Another option is to list the number of rules which match an endpoint.

**Comment by ianvernon** [comment]
> Couldn't this be confuse if we actually have the endpoint's Policy configuration showing enabled and then with the cilium endpoint list we would show disabled? 

When would this be the case?

**Comment by aanm** [comment]
>When would this be the case?

With 0 policies loaded, at least for k8s endpoints, `cilium endpoint config <ID>` will print `Policy Enabled` but it would show `False` in `cilium endpoint list`

**Comment by ianvernon** [comment]
I went ahead and created a script:
```
#!/bin/bash

TEST_NET="cilium"

function dump_ep_output {
        echo "-------- cilium endpoint list output --------"
        cilium endpoint list

        ENDPOINTS=`cilium endpoint list | grep 'ready' | awk '{print $1}'`
        for endpoint in $ENDPOINTS; do
                echo "-------- cilium endpoint config $endpoint --------"
                cilium endpoint config $endpoint
        done

        echo "-------- cilium config output --------"
        cilium config
}
function wait_ready {
        until [ "$(cilium endpoint list | grep ready -c)" -eq "3" ]; do
                echo "Waiting for all endpoints to be ready"
                sleep 2s
        done
}

function remove_containers {
        echo "-------- removing containers --------"
        ctrs=`docker ps -a | grep netperf | awk '{ print $1 }'`
        for ctr in $ctrs ; do
                docker rm -f $ctr
        done
}

remove_containers

echo "-------- starting containers --------"
docker run -dt --net=$TEST_NET --name foo -l id.foo -l id.teamA tgraf/netperf && docker run -dt --net=$TEST_NET --name bar -l id.bar -l id.teamA tgraf/netperf && docker run -dt --net=$TEST_NET --name baz -l id.baz tgraf/netperf

wait_ready
dump_ep_output

echo "-------- importing policy ---------"

cat <<EOF | cilium -D policy import -
[{
    "endpointSelector": {"matchLabels":{"id.bar":""}},
    "ingress": [{
        "fromEndpoints": [
            {"matchLabels":{"reserved:host":""}},
            {"matchLabels":{"id.foo":""}}
        ]
    }]
}]
EOF

wait_ready
dump_ep_output

echo "-------- deleting all policies --------"
cilium policy delete --all

wait_ready
dump_ep_output

remove_containers
```

The following runs of the script were with K8s *not* enabled.

Ex1 : `enable-policy true` 
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ ########################### true case ###########################
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ /tmp/policy_test.sh 
-------- removing containers --------
-------- starting containers --------
1e2a160667ff2b13ec90012761be9d4a4831b36c22e059783d4ec542e6506219
359f604f815f4336e0231249cf6f1c18bbd870a58b820d3d851244af379f0d79
d5e8fc4666cca95fcf13e09ad703a1744835b0878ad7306fd047466aa6ced01c
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Enabled       368        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Enabled       367        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Enabled       366        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Enabled
-------- cilium endpoint config 62006 --------
...
Policy                   Enabled
-------- cilium endpoint config 63464 --------
...
Policy                   Enabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- importing policy ---------
DEBU[0000] Entering directory -...                      
DEBU[0000] Loading file -                               
DEBU[0000] Constructed policy object for import [0xc42005eb40] 
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Enabled       368        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Enabled       367        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Enabled       366        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Enabled
-------- cilium endpoint config 62006 --------
...
Policy                   Enabled
-------- cilium endpoint config 63464 --------
...
Policy                   Enabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- deleting all policies --------
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Enabled       368        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Enabled       367        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Enabled       366        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Enabled
-------- cilium endpoint config 62006 --------
...
Policy                   Enabled
-------- cilium endpoint config 63464 --------
...
Policy                   Enabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- removing containers --------
d5e8fc4666cc
359f604f815f
1e2a160667ff
```

Ex2: `enable-policy false`
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ ########################### false case ###########################
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ /tmp/policy_test.sh 
-------- removing containers --------
-------- starting containers --------
896bd9b1d3f4c669bc484de2be65d736ac1b807f5d35be3b21a828ecf9d3a34e
5e9f2351ecdc531311a7e317b1d45a67c5d404804bce9278582bbbbdb6c4df59
6934fb2b739fbb7880d177b80de10559d3e2a77e8506d740106082fae7d7e82b
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Disabled      362        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Disabled      361        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Disabled      360        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Disabled
-------- cilium endpoint config 62006 --------
...
Policy                   Disabled
-------- cilium endpoint config 63464 --------
...
Policy                   Disabled
-------- cilium config output --------
...
Policy                   Disabled
PolicyTracing            Disabled
-------- importing policy ---------
DEBU[0000] Entering directory -...                      
DEBU[0000] Loading file -                               
DEBU[0000] Constructed policy object for import [0xc42005eba0] 
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Disabled      362        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Disabled      361        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Disabled      360        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Disabled
-------- cilium endpoint config 62006 --------
...
Policy                   Disabled
-------- cilium endpoint config 63464 --------
...
Policy                   Disabled
-------- cilium config output --------
...
Policy                   Disabled
PolicyTracing            Disabled
-------- deleting all policies --------
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
56326      Disabled      362        cilium:id.baz                 fd02::c0a8:210b:0:dc06   10.11.129.91    ready    
62006      Disabled      361        cilium:id.bar                 fd02::c0a8:210b:0:f236   10.11.116.202   ready    
                                    cilium:id.teamA                                                                 
63464      Disabled      360        cilium:id.foo                 fd02::c0a8:210b:0:f7e8   10.11.138.214   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 56326 --------
...
Policy                   Disabled
-------- cilium endpoint config 62006 --------
...
Policy                   Disabled
-------- cilium endpoint config 63464 --------
...
Policy                   Disabled
-------- cilium config output --------
...
Policy                   Disabled
PolicyTracing            Disabled
-------- removing containers --------
6934fb2b739f
5e9f2351ecdc
896bd9b1d3f4
```

Ex3: `enable-policy default`
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ ########################### default case ###########################
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ /tmp/policy_test.sh 
-------- removing containers --------
-------- starting containers --------
1819f52a08dba307f50b103b2850c97db2217df70682e5797442617d2fde3f48
d8e2c151aec5bb12ae6b520d8db63d41e3d102d307b2ec7688cc5a5e7e635528
750e0fd0288d9a75035c6ba95cfa8c0cec68dabd4e8a5a7abad1e6cb1afada0b
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
4314       Disabled      357        cilium:id.foo                 fd02::c0a8:210b:0:10da   10.11.25.58     ready    
                                    cilium:id.teamA                                                                 
4432       Disabled      359        cilium:id.baz                 fd02::c0a8:210b:0:1150   10.11.202.84    ready    
10687      Disabled      358        cilium:id.bar                 fd02::c0a8:210b:0:29bf   10.11.129.219   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 4314 --------
...
Policy                   Disabled
-------- cilium endpoint config 4432 --------
...
Policy                   Disabled
-------- cilium endpoint config 10687 --------
...
Policy                   Disabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- importing policy ---------
DEBU[0000] Entering directory -...                      
DEBU[0000] Loading file -                               
DEBU[0000] Constructed policy object for import [0xc42027a600] 
Waiting for all endpoints to be ready
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
4314       Enabled       357        cilium:id.foo                 fd02::c0a8:210b:0:10da   10.11.25.58     ready    
                                    cilium:id.teamA                                                                 
4432       Enabled       359        cilium:id.baz                 fd02::c0a8:210b:0:1150   10.11.202.84    ready    
10687      Enabled       358        cilium:id.bar                 fd02::c0a8:210b:0:29bf   10.11.129.219   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 4314 --------
...
Policy                   Enabled
-------- cilium endpoint config 4432 --------
...
Policy                   Enabled
-------- cilium endpoint config 10687 --------
...
Policy                   Enabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- deleting all policies --------
Waiting for all endpoints to be ready
-------- cilium endpoint list output --------
ENDPOINT   POLICY        IDENTITY   LABELS (source:key[=value])   IPv6                     IPv4            STATUS   
           ENFORCEMENT                                                                                              
4314       Disabled      357        cilium:id.foo                 fd02::c0a8:210b:0:10da   10.11.25.58     ready    
                                    cilium:id.teamA                                                                 
4432       Disabled      359        cilium:id.baz                 fd02::c0a8:210b:0:1150   10.11.202.84    ready    
10687      Disabled      358        cilium:id.bar                 fd02::c0a8:210b:0:29bf   10.11.129.219   ready    
                                    cilium:id.teamA                                                                 
-------- cilium endpoint config 4314 --------
...
Policy                   Disabled
-------- cilium endpoint config 4432 --------
...
Policy                   Disabled
-------- cilium endpoint config 10687 --------
...
Policy                   Disabled
-------- cilium config output --------
...
Policy                   Enabled
PolicyTracing            Disabled
-------- removing containers --------
750e0fd0288d
d8e2c151aec5
1819f52a08db
```

**Comment by ianvernon** [comment]
It seems like this will also require a change to `cilium config`, as it has the following output:
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium config 
Conntrack                Enabled
ConntrackAccounting      Enabled
ConntrackLocal           Disabled
Debug                    Enabled
DropNotification         Enabled
Policy                   Enabled
PolicyTracing            Disabled
```

We can change this with `cilium config Policy={Enabled,Disabled}` to enable / disable policy enforcement at the level of the daemon, but cannot set to `default` without relaunching the daemon. Should this be changed as well?

**Comment by tgraf** [comment]
> We can change this with cilium config Policy={Enabled,Disabled} to enable / disable policy enforcement at the level of the daemon, but cannot set to default without relaunching the daemon. Should this be changed as well?

As you note yourself, this would be a bigger change. Given that we can restart the daemon without losing state. I'm also fine if we only support this via flag and only support `enabled`=`default` and `disabled`=`never` via the API.

**Comment by tgraf** [comment]
Resolved by #815
